sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala

```diff
@@ -2019,14 +2008,27 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
         "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
   }
 
-  def cannotWriteNotEnoughColumnsToTableError(
-      tableName: String, expected: Seq[Attribute], query: LogicalPlan): Throwable = {
+  def insertMismatchedColumnNumberError(
+      firstNumColumns: Int,
+      invalidOrdinalNum: Int,
+      invalidNumColumns: Int): Throwable = {
+
+    def ordinalNumber(i: Int): String = i match {
+      case 0 => "first"
+      case 1 => "second"
+      case 2 => "third"
+      case i => s"${i + 1}th"
+    }
+
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1203",
+      errorClass = "NUM_COLUMNS_MISMATCH",
       messageParameters = Map(
-        "tableName" -> tableName,
-        "tableColumns" -> expected.map(c => s"'${c.name}'").mkString(", "),
-        "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
+        "operator" -> "INSERT INTO",
+        "firstNumColumns" -> firstNumColumns.toString,
+        "invalidOrdinalNum" -> ordinalNumber(invalidOrdinalNum + 1),
```

Comment:
To get `first`, `invalidOrdinalNum` should be -1?

Reply:
This is really weird. Let me modify it more reasonably.